pkg/controller/machineset/machineset_controller.go

```diff
@@ -230,12 +245,188 @@ func (c *Controller) processNextWorkItem() bool {
+		return nil
+	}*/
+
+	machinePrefix := ms.ObjectMeta.Name + "-"
+	//c.logger.WithField("machineset", ms.ObjectMeta.Name).Debugf("JDIAZ machine prefix: %v", machinePrefix)
+
+	machineSetMachines := []*clusteroperator.Machine{}
+	//machinesToCreate := []*clusteroperator.Machine{}
+	machinesToDelete := []*clusteroperator.Machine{}
+	//machinesToUpdate := []*clusteroperator.Machine{}
+
+	deleteCount := int(0)
+	createCount := int(0)
+
+
+	// caculate how many to create/delete
+	if len(machines) > ms.Spec.Size {
+		deleteCount = len(machines) - ms.Spec.Size
+	}
+	if len(machines) < ms.Spec.Size {
+		createCount = ms.Spec.Size - len(machines)
+	}
+
+	for _, machine := range machines {
+		if ! strings.HasPrefix(machine.ObjectMeta.Name, machinePrefix) {
+			// poorly-named
+			c.logger.WithField("machineset", ms.ObjectMeta.Name).Infof("JDIAZ child machine without expected prefix %v", machine.ObjectMeta.Name)
+			machinesToDelete = append(machinesToDelete, machine)
+			createCount++
+			deleteCount++
+			continue
+		}
+
+		if currentMachineNeedsUpdate(machine, ms) {
+			c.logger.WithField("machineset", ms.ObjectMeta.Name).Errorf("JDIAZ child machine needs update %v", machine.ObjectMeta.Name)
+			machinesToDelete = append(machinesToDelete, machine)
+			createCount++
+			deleteCount++
+			continue
+		}
+
+		machineSetMachines = append(machineSetMachines, machine)
+
+	}
+
+	// Sync machines
+	machineConfig := buildNewMachine(ms, machinePrefix)
+	c.logger.WithField("machineset", ms.ObjectMeta.Name).Debugf("JDIAZ machine template: %+v", machineConfig)
+
+	for i := 0 ; i < createCount ; i++ {
+		_, err := c.client.ClusteroperatorV1alpha1().Machines(ms.ObjectMeta.Namespace).Create(machineConfig)
+		if err != nil {
+			// FIXME add to error list?
+			c.logger.WithField("machineset", ms.ObjectMeta.Name).Error("JDIAZ error creating machine...")
+			// update expectations?
+		}
+	}
+
+	c.logger.WithField("machineset", ms.ObjectMeta.Name).Debug("JDIAZ delete count :v", deleteCount)
+	for _, m := range machinesToDelete {
+		c.logger.WithField("machineset", ms.ObjectMeta.Name).Debugf("JDIAZ deleting machine %v", m.Name)
+		c.client.ClusteroperatorV1alpha1().Machines(ms.ObjectMeta.Namespace).Delete(m.Name, &metav1.DeleteOptions{})
+		// catch error ^^^
+		deleteCount--
+	}
+
+	var machineToDelete *clusteroperator.Machine
+	c.logger.WithField("machineset", ms.ObjectMeta.Name).Debugf("JDIAZ deleting %v more machines", deleteCount)
+	for i := 0 ; i < deleteCount ; i++ {
+		machineToDelete, machineSetMachines = machineSetMachines[0], machineSetMachines[1:]
+		c.client.ClusteroperatorV1alpha1().Machines(ms.ObjectMeta.Namespace).Delete(machineToDelete.Name, &metav1.DeleteOptions{})
+		// catch error ^^^
+	}
+
+	return nil
+}
+
+func currentMachineNeedsUpdate(machine *clusteroperator.Machine, ms *clusteroperator.MachineSet) bool {
+	// compare machine settings with machineset settings to see if we have an out-of-date machine
+	if machine.Spec.NodeType != ms.Spec.NodeType {
```

Comment:
I would hope that this doesn't happen. That would imply that either a machine got attached to the wrong machineset or the node type of a machineset changed.

Reply:
is the nodetype field on a machineset expected to be mutable? if 'no', then perhaps there is no need for currentMachineNeedsUpdate() at all.

Reply:
Blocking that in validation seems fair to me, you shouldn't be able to change the node type of a machineset.